src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/JsonContent.cs

```diff
@@ -92,7 +110,20 @@ private async Task SerializeToStreamAsyncCore(Stream targetStream, CancellationT
             }
             else
             {
-                await JsonSerializer.SerializeAsync(targetStream, Value, ObjectType, _jsonSerializerOptions, cancellationToken).ConfigureAwait(false);
+                if (async)
+                {
+                    await JsonSerializer.SerializeAsync(targetStream, Value, ObjectType, _jsonSerializerOptions, cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+#if NETCOREAPP
+                    // Have to use Utf8JsonWriter because JsonSerializer doesn't support sync serialization into stream directly.
+                    using Utf8JsonWriter writer = new Utf8JsonWriter(targetStream);
+                    JsonSerializer.Serialize(writer, Value, ObjectType, _jsonSerializerOptions);
+#else
+                    throw new NotSupportedException(SR.SyncNotSupported);
```

Comment:
Why throw here but `Debug.Assert(async)` above? How can you trigger this error? Is there a test for it?

Reply:
Replaced with `Assert`, this should never be hit. `async = false` is passed only from `netcoreapp` compiled source.